### PR TITLE
Added initial Gabriel Benchmarks

### PR DESCRIPTION
**Pull request by Izaakwltn**
Here are the first few Gabriel Benchmarks implemented in Coalton (and lisp).

I also added a README with the basic running process.

### CODE REVIEW

File: benchmarks/gabriel-benchmarks/stak.lisp [hunk 1]
```diff
@@ -0,0 +1,75 @@
+;;;; gabriel-benchmarks/stak.lisp
+;;;;
+;;;;
+
+(cl:in-package #:coalton-benchmarks)
```
Comment (eliaslfox) [inline benchmarks/gabriel-benchmarks/stak.lisp:5]: `in-package`

File: benchmarks/gabriel-benchmarks/stak.lisp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+(define-benchmark stak ()
+  (declare (optimize speed))
+  (loop :repeat 1000
+        :do (with-benchmark-sampling
+              (coalton:coalton (coalton-benchmarks/native:stak 18 12 6))))
```
Comment (eliaslfox) [inline benchmarks/gabriel-benchmarks/stak.lisp:11]: Add type annotations and then you can call the function directly.

File: coalton.asd [hunk 3]
```diff
@@ -232,7 +232,10 @@
   :serial t
   :components ((:file "package")
                (:file "fibonacci")
-               (:file "big-float")))
+               (:file "big-float")
+               (:module "gabriel-benchmarks"
+                        :serial t
+                        :components ((:file "tak")))))
```
Comment (eliaslfox) [inline coalton.asd:238]: I think you're missing a couple files

**Review by eliaslfox — APPROVED** [review]
Thanks for porting these over

File: benchmarks/gabriel-benchmarks/stak.lisp (hunk 2 above)
Comment (Izaakwltn) [reply benchmarks/gabriel-benchmarks/stak.lisp:11]: Added!

File: benchmarks/gabriel-benchmarks/takr.lisp [hunk 4]
```diff
@@ -0,0 +1,1041 @@
...
+;;;
+;;;
+;;;
+
+
+(cl:in-package #:coalton-benchmarks/native)
+
+(cl:declaim (cl:optimize (cl:speed 3) (cl:safety 0)))
+
+(coalton-toplevel
```
Comment (eliaslfox) [inline benchmarks/gabriel-benchmarks/takr.lisp:536]: To be a reasonable benchmark every coalton function in this file should have type annotations.

**Review by eliaslfox — COMMENTED** [review]
The lisp functions should also have type annotations to keep the benchmarks fair

File: benchmarks/gabriel-benchmarks/takr.lisp (hunk 4 above)
Comment (Izaakwltn) [reply benchmarks/gabriel-benchmarks/takr.lisp:536]: Sounds good, this afternoon's zen garden

Comment (Izaakwltn) [reply benchmarks/gabriel-benchmarks/takr.lisp:734]: I added type annotations to every coalton and lisp function

File: benchmarks/gabriel-benchmarks/takl.lisp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+  (declare listn (UFix -> (List UFix)))
+  (define (listn n)
+    (if (not (== n 0))
+        (Cons n (listn (1- n)))
+        Nil))
+
+  (declare shorterp ((List UFix) -> (List UFix) -> Boolean))
+  (define (shorterp x y)
+    (and (not (coalton-library/list:null? y))
```
Comment (eliaslfox) [inline benchmarks/gabriel-benchmarks/takl.lisp:70]: `list:null?`

**Pull request merged**